### PR TITLE
feat(react-persona): Adding initial implementation

### LINKED ISSUE #24640: Persona: Add initial implementation

### PR DESCRIPTION
**Pull request by sopranopillow**
This PR adds the initial implementation of Persona. 

Note: Styles are still a WIP, this PR is mainly to get the logic down.

## Related Issue(s)

Fixes #24640

### CODE REVIEW

**Comment by fabricteam** [comment]
## 📊 Bundle size report

<sub>🤖 This report was generated against <a href='https://github.com/microsoft/fluentui/commit/7f8bad094dadcdb671d5746848e96b5bd4b5791a'>7f8bad094dadcdb671d5746848e96b5bd4b5791a</a></sub>
<div id="bundleSizeComment9423"/>

File: packages/react-components/react-persona/src/components/Persona/usePersonaStyles.ts [hunk 1]
```diff
@@ -1,33 +1,232 @@
...
+});
+
+const useGridStyles = makeStyles({
+  after: {
+    gridTemplateAreas: `
+    "${COIN_GRID_AREA} ${PRIMARY_TEXT_GRID_AREA}"
+    "${COIN_GRID_AREA} ${SECONDARY_TEXT_GRID_AREA}"
+    "${COIN_GRID_AREA} ${TERTIARY_TEXT_GRID_AREA}"
+    "${COIN_GRID_AREA} ${QUATERNARY_TEXT_GRID_AREA}"
+    `,
```
Comment (behowell) [inline packages/react-components/react-persona/src/components/Persona/usePersonaStyles.ts:40]: Instead of using `gridTemplateAreas`, you can rely on grid's auto-flow behavior, and the DOM order.

* By default, root only needs `display: 'inline-grid'`.
* If `textPosition !== 'below'`
  * Set the root's `gridAutoFlow: 'column'` (the default grid flow of 'row' works for text below).
  * Set the avatar/presence slots' `gridRowStart: 'span 5'` (5 rows lets it extend one row past the text lines, so a large avatar doesn't mess up the height of the rows for the text lines)
* If `textPosition === 'before'`
  * When rendering, re-order the DOM to put the text slots before the avatar/presence badge slot.

See an example here: https://codepen.io/behowell/pen/RwyoXqz?editors=1100

File: packages/react-components/react-persona/src/components/Persona/usePersonaStyles.ts [hunk 2]
```diff
@@ -1,33 +1,232 @@
...
+      }
+    }
+  } else {
+    if (numTextLines > 1) {
+      primaryTextClassNames.push(textStyles.body1);
+    } else {
+      primaryTextClassNames.push(textStyles.caption1);
+    }
+    optionalTextClassNames.push(textStyles.caption1);
+  }
```
Comment (behowell) [inline packages/react-components/react-persona/src/components/Persona/usePersonaStyles.ts:226]: By pushing these classes to an array, it seems to imply that they could get more than one size class, which I'm assuming isn't the case? (It's a little hard to follow all of the branches here).

Could these just be variables you assign like `primaryTextSizeClass = textStyles.body1`, so that it's clear there will be only one? Then the return value of this function could be:
```ts
  return {
    primaryTextClassName: mergeClasses(textStyles.base, textStyles.primaryText, primaryTextSizeClass),
    optionalTextClassName: mergeClasses(textStyles.base, textStyles.optionalText, optionalTextSizeClass),
  };
```

File: packages/react-components/react-persona/src/stories/Persona/PersonaDefault.stories.tsx [hunk 3]
```diff
@@ -1,4 +1,45 @@
...
+        presenceOnly
+        presence={{ status: 'out-of-office' }}
+        primaryText="PrimaryText"
+        secondaryText="Secondary text"
+        tertiaryText="Tertiary text"
+        quaternaryText="Quaternary text"
+        {...props}
+      />
+    </div>
+  );
```
Comment (behowell) [inline packages/react-components/react-persona/src/stories/Persona/PersonaDefault.stories.tsx:44]: The Default story should be a single Persona that showcases its common features (e.g. maybe an avatar+presence, with primary and secondary text). The other stories can show examples of the other variants like `presenceOnly`, or with more lines of text, etc.

File: packages/react-components/react-persona/src/stories/Persona/PersonaAvatarSize.stories.tsx [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+});
+
+const sizes: AvatarSizes[] = [16, 20, 24, 28, 32, 36, 40, 48, 56, 64, 72, 96, 120, 128];
+
+export const AvatarSize = () => {
+  const styles = useStyles();
+
+  return (
+    <div className={styles.root}>
+      {sizes.map(size => (
```
Comment (behowell) [inline packages/react-components/react-persona/src/stories/Persona/PersonaAvatarSize.stories.tsx:20]: To keep the example code simpler, it's probably better to write out the example for each size (maybe just one example per size, with primary and secondary text). See AvatarSize.stories.tsx or PresenceBadgeSizes.stories.tsx.

Comment (behowell) [reply packages/react-components/react-persona/src/stories/Persona/PersonaAvatarSize.stories.tsx:20]: These size stories could also serve the purpose of showing how fixed sizing works. (Instead of separate stories for fixed sizing)

File: packages/react-components/react-persona/src/stories/Persona/PersonaAvatarSize.stories.tsx [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+            presence={{ status: 'out-of-office' }}
+            primaryText="PrimaryText"
+            secondaryText="Secondary text"
+          />
+          <Persona name="Kevin Sturgis" size={size} presence={{ status: 'out-of-office' }} primaryText="PrimaryText" />
+        </>
+      ))}
+    </div>
+  );
+};
```
Comment (behowell) [inline packages/react-components/react-persona/src/stories/Persona/PersonaAvatarSize.stories.tsx:51]: All of these examples should also have descriptions that how the feature works that they're demonstrating.

File: packages/react-components/react-persona/src/stories/Persona/PersonaSizingScaled.stories.tsx [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+      />
+      <Persona
+        sizing="scaled"
+        presenceOnly
+        presence={{ status: 'out-of-office' }}
+        primaryText="Primary text"
+        secondaryText="Secondary text"
+      />
+      <Persona sizing="scaled" presenceOnly presence={{ status: 'out-of-office' }} primaryText="Primary text" />
+    </div>
```
Comment (behowell) [inline packages/react-components/react-persona/src/stories/Persona/PersonaSizingScaled.stories.tsx:69]: Ideally this would be simplified to have just a few examples that show how scaled sizing works.

File: packages/react-components/react-persona/src/components/Persona/usePersona.ts [hunk 7]
```diff
@@ -12,17 +14,95 @@ import type { PersonaProps, PersonaState } from './Persona.types';
...
+    if (sizing === 'scaled') {
+      if (numTextLines === 1) {
+        state.presence.size = 'small';
+      } else if (numTextLines === 2) {
+        state.presence.size = 'large';
+      } else {
+        state.presence.size = 'extra-large';
+      }
+    }
+  }
```
Comment (behowell) [inline packages/react-components/react-persona/src/components/Persona/usePersona.ts:105]: These values shouldn't override user-specified values (like state.presence.size) if they were provided. One way to do it is to calculate the values ahead of time and pass them in `defaultProps` in `resolveShorthand`. Or at least check that the prop isn't set before overwriting it.

File: packages/react-components/react-persona/src/components/Persona/Persona.test.tsx [hunk 8]
```diff
@@ -1,18 +1,11 @@
...
     Component: Persona,
     displayName: 'Persona',
-  });
-
-  // TODO add more tests here, and create visual regression tests in /apps/vr-tests
-
-  it('renders a default state', () => {
-    const result = render(<Persona>Default Persona</Persona>);
-    expect(result.container).toMatchSnapshot();
+    disabledTests: ['component-has-static-classnames-object'],
```
[the remark this answers is not in the document]
Comment (sopranopillow) [reply packages/react-components/react-persona/src/components/Persona/Persona.test.tsx:8]: Here's the issue mentioning this #24859

**Review by behowell — APPROVED** [review]
Added some comments, but I'm going to approve it so you can merge after addressing them. Thanks!

File: packages/react-components/react-persona/src/components/Persona/usePersonaStyles.ts [hunk 9]
```diff
@@ -1,33 +1,192 @@
...
 
 /**
  * Styles for the root slot
  */
 const useStyles = makeStyles({
   root: {
-    // TODO Add default styles for the root element
+    display: 'inline-grid',
+    gridAutoColumns: 'max-content',
+    gridAutoRows: 'max-content',
```
[the remark this answers is not in the document]
Comment (sopranopillow) [reply packages/react-components/react-persona/src/components/Persona/usePersonaStyles.ts:23]: `gridAutoRows: max-content` is needed because when you have a larger height than the actual needed, it will stretch and mess up the spacing:
image

for the columns it's the same issue, but it's worse since the coin will stretch and the text will be off:
image
